src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,327 @@
-     * @param samRgRecords           a list of read groups
-     * @param HistogramMADTolerance  MAD tolerance when producing histogram plot
+     * Constructor; also acts as driver for organizing workflow.
+     *
+     * @param filteredReads         reads that pass filters
+     * @param header                header in the input
+     * @param accumLevels           accumulation level {ALL_READS, SAMPLE, LIBRARY, READ_GROUP}
+     * @param histogramMADTolerance MAD tolerance when producing histogram plot
      */
-    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
     public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
-                                           //final Set<MetricAccumulationLevel> accumLevels,
-                                           final List<SAMReadGroupRecord> samRgRecords,
-                                           final double HistogramMADTolerance) {
-
-        final JavaRDD<Integer> lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
-        lengthRDD.cache();
+                                           final SAMFileHeader header,
+                                           final Set<MetricAccumulationLevel> accumLevels,
+                                           final double histogramMADTolerance) {
+
+        /* General strategy:
+           construct untrimmed hand rolled "histogram" (SortedMap) of all read groups in three steps,
+             because htsjdk Histogram does not play well with Serialization
+           so first hand roll a histogram at the read-group level during the 1-pass all reads traversal on Spark.
+           when computing InsertSizeMetrics, first convert the SortedMap version histogram to htsjdk Histogram, then compute metrics.*/
+
+        final Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsOfReadGroups = filteredReads.mapToPair(read -> traverseReadsToExtractInfo(read, header))
+                                                                                                                                   .groupByKey()
+                                                                                                                                   .mapToPair(InsertSizeMetricsCollectorSpark::gatherSizesByOrientation)
+                                                                                                                                   .mapToPair(InsertSizeMetricsCollectorSpark::constructHistogramFromList)
+                                                                                                                                   .collectAsMap();
+
+        // accumulate for higher levels, if so desired
+        Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsOfLibraries = accumLevels.contains(MetricAccumulationLevel.LIBRARY  ) ? new HashMap<>() : null;
```

Comment:
This is a much cleaner way of doing things than before, thanks for re-writing. I have one more suggestion:

Rather than initializing these to either null or an empty map, and then checking whether the map is null in the aggregateHistograms() method, you could always initialize these maps to empty and just pass the set of desired levels (accumLevels) into the aggregate method, and in that method only add the appropriate extractors if the level is in the set. I think that conveying information by setting a reference to null tends to lead to confusing code and/or bugs, and in this case you already have the necessary information (the level set) so you might as well just pass that instead. We can discuss in person if you'd like.